Various smaller functionality improvements and cleanups

Wentzell opened the pull request:
C.f. separate commit messages

Wentzell committed:
Add factory functions nda::rand(...) and nda::basic_array<...>::rand + test

Wentzell committed:
Allow explicit value_type conversion when cross-constructing arrays

Wentzell committed:
Minor cleanups in nda_storage test

Wentzell committed:
Move allocators and handle into common 'mem' namespace and directory

Wentzell committed:
Undeprecate first_dim and second_dim functions, remove higher XYZ_dim functions

Wentzell committed:
Add missing header includes for improved language-server messages

Wentzell committed:
Minor doc and code improvements in matrix_functions.hpp and nda_zero.cpp

Wentzell committed:
Add numpy-like functions nda::diagonal and nda::diag + test

Wentzell committed:
Fix Bug in implementation of nda::dot when using complex conjugation

Wentzell committed:
Add array and view member function 'indices()' to return generator of all index tuples

Wentzell committed:
In nda::blas::dot allow scalars as arguments

Wentzell committed:
Fix Bug debug checks of nda::blas::ger implementation

Wentzell committed:
Add function nda::blas::outer_product and test

Wentzell committed:
Minor simplification in nda::diag implementation

Wentzell committed:
Allow array and view assignment from 1d contiguous ranges

Wentzell committed:
Allow basic_array_view construction from and value comparison with contiguous ranges

parcollet commented on c++/nda/basic_array.hpp:
```diff
@@ -117,8 +118,8 @@ namespace nda {
      */
-    template <std::integral Int, typename RHS>
-    explicit basic_array(Int i, RHS const &val) noexcept requires((std::is_integral_v<Int> and Rank == 1 and is_scalar_for_v<RHS, basic_array>)) {
+    template <std::integral Int = long, typename RHS>
```
why = long ? 
Isn't Int given by the first parameter ?

Wentzell replied:
Yes, the default can be removed here.

parcollet commented on c++/nda/basic_array.hpp:
```diff
@@ -261,6 +262,19 @@ namespace nda {
+      auto static dist = std::uniform_real_distribution<>(0.0, 1.0);
+
+      auto res = basic_array{shape};
+      auto l = [&res](auto const &...args) mutable { res(args...) = dist(gen); };
```
why not simply:
for (auto & x : res) x = dist(gen);
?
Just iterate on the array, and pick up random number.  No need of a complicated foreach ?

Wentzell replied:
Yes, I agree. Those lines were in fact copied from `fill_with_scalar` where the same simplification should apply!?
https://github.com/TRIQS/nda/blob/2d28f423b2439698aae0f1f00bcbbe04c8cdc499/c%2B%2B/nda/_impl_basic_array_view_common.hpp#L355-L356
Please confirm.

parcollet commented on c++/nda/basic_functions.hpp:
```diff
@@ -35,6 +35,29 @@ namespace nda {
 
+  // --------------------------- rand ------------------------
+
+  /// Create an array the given dimensions and populate it with random
```
doc is incorrect.
It returns a scalar for Rank 0

parcollet commented on c++/nda/basic_functions.hpp:
```diff
@@ -35,6 +35,29 @@ namespace nda {
+      auto static gen  = std::mt19937(std::random_device{}());
+      auto static dist = std::uniform_real_distribution<>(0.0, 1.0);
+      return dist(gen);
+;
```
? isolated ;

parcollet commented on c++/nda/basic_functions.hpp:
```diff
@@ -35,6 +35,29 @@ namespace nda {
+  }
+
+  ///
+  template <typename RealType = double, std::integral... Int>
```
rand<double>( 1,2,3) ?
I would prefer a more explicit name:
random_array ?
It is not clear when reading a code that rand returns and array ...

Wentzell replied:
As discussed: Names are chosen based on `numpy` function names, c.f. `diag`, `diagonal`, `zeros`, `rand`, ..

parcollet commented on test/c++/nda_storage.cpp:
```diff
@@ -23,19 +23,10 @@ using namespace nda::mem;
-};
-
-TEST_F(Ref, HR) { // NOLINT
+TEST(Storage, HR) { // NOLINT
```
I understand it is not used, 
but do we have a test checking that all memory is deallocated in test ? 
--> if not : add an issue to do it.

parcollet commented on c++/nda/basic_functions.hpp:
```diff
@@ -58,6 +58,22 @@ namespace nda {
     return rand<RealType>(std::array<long, sizeof...(Int)>{i...});
   }
 
+  // --------------------------- dim helpers ------------------------
```
Why this commit ? 
Why not replace first_/second_dim ? 
Isn't it strange to provide only these two ?
Or does it imply too many changes in old code ?

parcollet commented on c++/nda/matrix_functions.hpp:
```diff
@@ -36,10 +38,9 @@ namespace nda {
   }
 
-  ///
+  /// Return the conjugate transpase of a matrix or a rank==2 array
```
typo
transPOSE

parcollet commented on test/c++/nda_zero.cpp:
```diff
@@ -28,9 +28,9 @@ TEST(NDA, Zeros) { //NOLINT
-  auto a1 = nda::array<long, 1>::zeros(std::array{3});
-  auto a2 = nda::array<long, 2>::zeros(std::array{3, 4});
-  auto a3 = nda::array<long, 3>::zeros(std::array{3, 4, 5});
+  auto a1 = nda::array<long, 1>::zeros({3});
```
Should'n't zeros also accept
zeros(1,2,3) ??
Matter of taste...

parcollet commented on c++/nda/matrix_functions.hpp:
```diff
@@ -47,7 +47,23 @@ namespace nda {
+  template <ArrayOfRank<2> M>
+  ArrayOfRank<1> auto diagonal(M &m)  {
+    long dim = std::min(m.shape()[0], m.shape()[1]);
+    long stride = stdutil::sum(m.indexmap().strides());
```
indexmap().strides()[0] + indexmap().strides()[1]
might be clearer (not need to know sum), and faster (no loop !)
?

parcollet commented on c++/nda/matrix_functions.hpp:
```diff
@@ -47,7 +47,23 @@ namespace nda {
 
-  ///
+  /// Return a vector_view on the diagonal of a matrix or a rank==2 array
+  template <ArrayOfRank<2> M>
```
I think it is wrong.
You use the concept Array, which is not necessarily a container (could be an expression)
and then use .data() in the code.

Either : 
1-  one makes a view  on a container.
2- OR we make a lazy object with Array<1> concept, returning the a( j,j)

Both are useful
1- I would call it diagonal_view ? It is a view of data .... could indeed be const or not with the implementation below, almost.
2- diagonal, but implemented as 

template<typename M > 
struct diagonal_of_matrix { // name ?? 
  M m; // a view or anything ... Can not be const &, because of forward of temporaries.
  operator()(i) const { return M(i,j);}
  shape() const { return min ....}
};

Array<1> auto diagonal (ArrayOfRank<2> auto const & m) { return diagonal_of_matrix{m()};} // () is crucial h !

parcollet commented on c++/nda/matrix_functions.hpp:
```diff
@@ -47,7 +47,23 @@ namespace nda {
+  }
+
+  /// Return a new matrix with the values of v on the diagonal
+  template <ArrayOfRank<1> V>
```
We should use a better name here, more explicit ...
At the end, when reading a code `diagonal`, `diag` is not very explicit

parcollet commented on c++/nda/blas/dot.hpp:
```diff
@@ -54,15 +55,21 @@ namespace nda::blas {
   auto dot(X const &x, Y const &y) {
-    EXPECTS(x.shape() == y.shape());
-    return _dot_impl<false>(x, y);
+    if constexpr (Scalar<X> or Scalar<Y>) {
```
why ?
That is not the expected blas interface.
I am worried it can surprise people.

At minimum, I would write a clear doc 
or 
make 4 (or 2) implementations
dot(Scalar, Scalar)
dot(Array<1>, Array<1>)
etc...
to have a clear documentation with 4 overloads

parcollet replied:
Ps : 
or simply use ArrayOrScalar in the signature ...

parcollet commented on c++/nda/blas/ger.hpp:
```diff
@@ -47,8 +47,8 @@ namespace nda::blas {
 
-    EXPECTS(m.extent(1) == x.extent(0));
-    EXPECTS(m.extent(0) == y.extent(0));
+    EXPECTS(m.extent(0) == x.extent(0));
```
Does it that previously test where passing but only using square cases ??

Wentzell replied:
Yes the test was only checking square cases. I will adjust the outer_product case to cover a non-square case.

parcollet commented on c++/nda/blas/ger.hpp:
```diff
@@ -65,4 +65,36 @@ namespace nda::blas {
+   *
+   * @return The outer product of a and b as defined above
+   */
+  template <ArrayOrScalar A, ArrayOrScalar B>
```
Same problem as with diagonal.
ArrayOrScalar can be anything, not a view or a basic_array.
Then reshape_view does not work (and any xxx_view ...)

We probably need a concept
ArrayContainer<R> or alike
to be used only with containers...

TBD.

parcollet replied:
Otherwise the code looks fine

Wentzell replied:
Yes, I agree. I have added the `MemoryArray` concept now that covers this 


```c++
template <typename A>
concept MemoryArray = Array<A> && requires(A &a) {

   // We can acquire the pointer to the underlying data
   { a.data() } -> std::same_as<std::conditional_t<std::is_const_v<A>, const get_value_t<A>, get_value_t<A>> *>;

   // Exposes the memory stride for each dimension
   { a.indexmap().strides() } -> StdArrayOfLong;
};

parcollet commented on test/c++/nda_blas.cpp:
```diff
@@ -123,6 +120,21 @@ TEST(BLAS, ger) { //NOLINT
+TEST(BLAS, outer_product) { //NOLINT
+
+  auto N = nda::rand<double>(2, 2);
+  auto M = nda::rand<double>(2, 2);
```
Put numbers in the test ! rand ?
The test does not test so much as everything is 0, just the same.
one could change k and l and it would pass, no ?

Wentzell replied:
I was already using `rand` here!? Why are you saying that everything is 0?

parcollet commented on c++/nda/_impl_basic_array_view_common.hpp:
```diff
@@ -307,9 +307,9 @@ auto &operator/=(RHS const &rhs) noexcept {
-template <typename C>
-requires(Rank == 1 and std::ranges::contiguous_range<C>) auto &operator=(C const &rhs) noexcept {
-  *this = array_const_view<typename C::value_type, 1, C_layout>{{long(rhs.size())}, std::to_address(std::begin(rhs))};
+template <typename R>
```
I am concerned by this, for the reason you mentioned.
In the future, we may simply want an array<T, 1> to be a contiguous_range. Because it is, no ?
Then there may be a collision here.
We would : 
- add the ArrayContainer concept --> cf previous commit, TBD
- require !ArrayContainer : i.e. any range but a container, which is treated a different way
- (it should not be necessary because of template rules, but still a lot clearer)
-
TBD.

parcollet commented on c++/nda/basic_array_view.hpp:
```diff
@@ -147,6 +147,15 @@ namespace nda {
+     * @tparam R The contiguous Range type
+     * @param r The contiguous Range
+     */
+    template <std::ranges::contiguous_range R>
```
Same as before

Wentzell committed:
Add factory functions nda::basic_array<..>::ones and nda::ones + test

- Minor doc improvements for factory functions 'ones', 'zeros', 'rand'
- Bugfix in nda::zeros(..): Fix zero initialization for structs with
  multiple members

Wentzell committed:
Remove unnecessary default in template parameter of basic_array(Int i, RHS const &val)

Wentzell committed:
Add MemoryArray concept and extend Array concept to include size() function

-MemoryArray makes sure we can access the underlying data and layout in memory

Wentzell committed:
Incorporate smaller PR review comments

hmenke commented on test/c++/nda_basic.cpp:
```diff
@@ -22,6 +22,13 @@ static_assert(nda::is_scalar_for_v<int, matrix<std::complex<double>>> == 1, "oop
   nda::array<long, 2> A(3, 3);
   EXPECT_EQ(A.shape(), (nda::shape_t<2>{3, 3}));
+
+  nda::array<long, 2> B({3, 3});
```
This doesn't work on GCC 10.
```c++
/tmp/nda/test/c++/nda_basic.cpp:26:31: error: converting to ‘nda::basic_array<long int, 2, nda::C_layout, 'A', nda::heap>’ from initializer list would use explicit constructor ‘nda::basic_array<ValueType, Rank, Layout, Algebra, ContainerPolicy>::basic_array(Int ...) [with Int = {int, int}; ValueType = long int; int Rank = 2; Layout = nda::C_layout; char Algebra = 'A'; ContainerPolicy = nda::heap]’
   26 |   nda::array<long, 2> B({3, 3});
      |   
```

hmenke replied:
I think this was supposed to be
```suggestion
  nda::array<long, 2> B(std::array<long, 2>{3, 3});
```
but the bare initialization list `{3, 3}` cannot be implicitly converted here because the constructor from `std::array` is marked `explicit`.

Wentzell replied:
Yes, it's the right thing to construct the std::array here to avoid confusion with the initializer list constructors.
This should be fixed now. It's still strange that this compiles with clang and not gcc.
